Disable URL Metric storage locking by default for administrators

westonruter opened the pull request:
<!-- Please reference the issue this PR fixes. If this PR does not fix the entire issue, change this to Addresses #... instead. -->
Fixes #1829

This changes the default value passed into the `od_url_metric_storage_lock_ttl` from `60` to `0` when the current user is an administrator. Additionally, a new `od_store_url_metric_now` meta capability is introduced which maps to `manage_options` by default. This gives sites the ability to customize which users have access to that capability.

In practice this will eliminate the need for the following code in the [Optimization Detective Dev Mode plugin](https://github.com/westonruter/od-dev-mode) as long as the developer is testing while being logged-in as an administrator:

```php
// During development, this can be useful to set to zero so that you don't have to wait for new URL Metrics to be requested when engineering a new optimization.
add_filter(
	'od_url_metric_freshness_ttl',
	static function () {
		return 0;
	}
);
```

westonruter committed:
Disable storage locking for administrators

westonruter committed:
Introduce od_store_url_metric_now meta cap which maps to manage_options by default

felixarntz commented:
@westonruter I'm not sure about this change. This seems to be something that helps development, so I think having it in the OD Dev Mode plugin makes more sense to me than making it the default for any administrator. Many administrators of production sites are probably not developers.

On another note, I think the code snippet referenced in the PR description is for something else than described (it mentions viewport, rather than storage lock)?

westonruter commented:
> On another note, I think the code snippet referenced in the PR description is for something else than described (it mentions viewport, rather than storage lock)?

Oops. I copied the wrong code. This has been corrected.

> I'm not sure about this change. This seems to be something that helps development, so I think having it in the OD Dev Mode plugin makes more sense to me than making it the default for any administrator. Many administrators of production sites are probably not developers.

Actually, it's not specifically to help development. If someone creates a new post and then views it on the frontend, this should result in a URL Metric being collected right away without having to wait for any TTL storage lock to expire for the author or for another visitor to come by to view the post. A post author will almost always want to view the published post on the frontend after publishing, so this will result in a URL Metric being stored right away to give the collection a head start. Additionally, an administrator browsing around their site should be able to submit URL Metrics as they go around without having to wait for the storage lock TTL to expire. The purpose of the storage lock is to guard against unlimited DB writes from a frontend visitor. This, however, is not relevant to administrators (or even other roles above subscriber) since they can write to the DB at will in the admin.

Actually, come to think of it, `manage_options` probably isn't the right capability here, now that I think about it. In reality it should be `edit_posts`. It's about whether or not a user has "write access" to the database.

felixarntz commented:
Your rationale makes sense to me. In that case, maybe a better implementation to achieve that goal would be to invalidate the current user's storage lock transient whenever a post is updated? That way, any fresh content would be able to receive updates immediately, without e.g. "blasting" the server with tons of URL metrics (in case of a very unpopular site, where the administrator visits e.g. 15 URLs quickly to test stuff).

westonruter commented:
I'm not sure that would work. The storage lock TTL is sent to the client to prevent the client from even attempting to make a `POST` request to store a new URL Metric. 

I don't see a problem with allowing an administrator to make multiple DB writes quickly. Consider bulk actions in the admin or quickly moderating through comments, for example.

westonruter commented:
Oh, I just realized this won't work as-is since the `_wpnonce` isn't being included for logged-in users, so the endpoint at present would always see the user as anonymous. Easy to fix.

westonruter committed:
Include REST API nonce in URL Metric storage requests when the user is logged-in

westonruter commented:
> Oh, I just realized this won't work as-is since the `_wpnonce` isn't being included for logged-in users, so the endpoint at present would always see the user as anonymous. Easy to fix.

There, this is fixed by 9f89cd4. Note that this could eventually mitigate the need for the Site Health test to check whether anonymous REST API requests are allowed (#1731). We could potentially add a new hook which disables URL Metric collection for anonymous users. This could be attractive to high-traffic sites (cf. #1655) which don't want to expose the REST API to anonymous visitors. Such sites would require logged-in users to collect the URL Metrics by browsing around, but also some automated collection could be implemented in the background, such as by opening the frontend in an `IFRAME` in the background when saving the post (https://github.com/WordPress/performance/issues/1311).

westonruter committed:
Add missing test for filter_map_meta_cap

westonruter commented:
@felixarntz Anything else you want to see here?

felixarntz commented:
@westonruter Sorry I didn't get to responding back to this sooner.

> I'm not sure that would work. The storage lock TTL is sent to the client to prevent the client from even attempting to make a `POST` request to store a new URL Metric. 

But if the storage lock was cleared, the lock wouldn't be there anymore until they send a URL metric again, correct? You mentioned previously as a rationale for this change:

> If someone creates a new post and then views it on the frontend, this should result in a URL Metric being collected right away without having to wait for any TTL storage lock to expire for the author or for another visitor to come by to view the post.

For that reason, I suggested to clear the TTL for the current user when a post is updated. If they then visit the post in the frontend, URL metric data for that post would be sent, no? So I'm thinking that would satisfy this use-case without making broader changes as proposed here.

> I don't see a problem with allowing an administrator to make multiple DB writes quickly. Consider bulk actions in the admin or quickly moderating through comments, for example.

To me it's less about "allowing" than about the implications on low-quality servers. The examples you're mentioning are still actual actions taking by the user, so there's naturally more of a limitation on how fast the user can perform those. There's a slightly different implication on simply visiting pages, and the administrator (in many cases probably) not knowing that each page load will be a database write.

FWIW I'm not strongly opposed to this change, but based on what you described as rationale, I'm thinking there may be a less risky way to make it possible, as outlined above.

westonruter commented:
Storage locks are not bound to specific users. They are bound to a hash of the IP address. So if that lock were to be cleared, then anyone from that IP address could then submit a URL Metric, regardless of whether they are an admin or not. 

Even in the case of an admin, however, clearing the storage lock wouldn't necessarily help because the storage lock TTL is still 60 seconds. This lock is stored client-side in `sessionStorage` to circumvent the user from needlessly attempting `POST` requests when they should be locked.

Lastly, if the storage lock were only cleared in response to a post update in the DB, then this wouldn't address the goal of an administrator being able to cause URL Metrics to be collected just by browsing the site. Let's say a user goes to their site, goes to create a post and publishes it. Once they do so and view the post on the frontend, a URL Metric should be collected. If the storage lock were cleared in response to editing a post, then good, the URL Metric should successfully be collected. But what if they then go to the homepage to see that post showing up there? The URL Metric wouldn't be collected because the storage lock would be re-imposed. If, however, the storage lock is disabled by default for admins, then URL Metrics will be collected as needed while navigating around the site without restriction. Given that DB writes happen all the time for admin users, I don't see this as an issue. DB writes happen for admins just by visiting the New Post screen and then they happen in the background for autosaves for that screen too.

felixarntz commented:
> Storage locks are not bound to specific users. They are bound to a hash of the IP address. So if that lock were to be cleared, then anyone from that IP address could then submit a URL Metric, regardless of whether they are an admin or not.

Good point, I didn't consider that - of course not guaranteed that the IP address is the same person. Fair point also on the `sessionStorage` consideration.

> Lastly, if the storage lock were only cleared in response to a post update in the DB, then this wouldn't address the goal of an administrator being able to cause URL Metrics to be collected just by browsing the site.

I see how that can be valuable, but I'm still wary of unexpected side effects on scalability, depending on a few factors:
* The administrator sending a database write for more or less every URL they view on the site is probably fine, as it shouldn't have performance implications. Likely 90+% of WordPress sites have 1 administrator, so that'd be okay.
* You mentioned above though that you would like to change the capability required to `edit_posts`. In that scenario, the implications would change - I would specifically become worried about e.g. certain membership site setups where it wouldn't be too far fetched to give 1000s of users `edit_posts`. Probably still would be a specific scenario, but in any case, it certainly changes the impact that this change would have.
* There is a question on the diversity of URL metrics. Does it matter or not? Maybe it's fine for the vast majority of URL metrics to come from a single user - the viewport implications for the same viewport shouldn't differ wildly. But I want to make sure we consider that before going the route. Because effectively, for sites with very few to no users, implementing this change might lead to a similar result as if we had e.g. some admin UI to automatically visit all URLs as the authenticated user (which we discussed before, but never went for).

westonruter committed:
Merge branch 'trunk' into remove/storage-lock-for-admins

westonruter commented:
Good points. 

> * You mentioned above though that you would like to change the capability required to `edit_posts`. In that scenario, the implications would change - I would specifically become worried about e.g. certain membership site setups where it wouldn't be too far fetched to give 1000s of users `edit_posts`. Probably still would be a specific scenario, but in any case, it certainly changes the impact that this change would have.

Yeah, I withdraw my idea to have the capability be `edit_posts` by default for that reason as well.

> * There is a question on the diversity of URL metrics. Does it matter or not? Maybe it's fine for the vast majority of URL metrics to come from a single user - the viewport implications for the same viewport shouldn't differ wildly. But I want to make sure we consider that before going the route. Because effectively, for sites with very few to no users, implementing this change might lead to a similar result as if we had e.g. some admin UI to automatically visit all URLs as the authenticated user (which we discussed before, but never went for).

This is something I thought about too, where it would be nice if a given user only was allowed to contribute one fresh URL Metric for a given URL. However, in order to do this, we'd have to store some identifier about the user session in the URL Metric to know whether or not we should reject additional URL Metrics from that client. This would be problematic from a privacy perspective, since then the URL Metrics would essentially become a way to track users. 

We could possibly have a flag in the URL Metric for whether it was was collected from an authenticated user or not (e.g. `userAuthenticated` boolean property), which could be useful as a way possibly to differentiate trusted vs untrusted URL Metrics in the future. This flag could also be used to try to ensure that there is only one such fresh `userAuthenticated` URL Metric collected at a time. But what if it is a membership site in which all users are authenticated? That won't work then. Could we store the authenticated `userId` in the URL Metric and only allow one fresh URL Metric to be stored at a given time from that user (if it is not zero or `null`)? This goes back to the privacy risk.

For the sake of diversity of URL Metric collection, if an anonymous user visits the homepage, clicks to view a post for a minute, then goes back to the homepage, and then reads another post for a minute, before going back to the homepage for a third time: this will result in 3 URL Metrics being collected from that same user. Ideally we wouldn't collect multiple URL Metrics from the same user, but this isn't being thwarted currently anyway. I think it's more important for URL Metrics to be collected from the same user over the alternative of not collecting any URL Metric at all.

Nevertheless, here's an alternative approach extending on top of what this PR currently implements. What if we use `sessionStorage` to keep track of the URLs which the client had submitted URL Metrics for? When `detect.js` boots up it can check if the `currentUrl` was already submitted, and if so, short-circuit. This addresses the privacy concern of storing some identifier about the user session with the URL Metric while also ensuring a diversity of clients being the source for the submitted URL Metrics. SG?

westonruter committed:
Prevent submitting duplicate URL Metrics from one client session

westonruter commented:
@felixarntz There, what do you think about a57e080? The hashing of the URL in the `sessionStorage` may be overkill.

felixarntz commented:
> > * There is a question on the diversity of URL metrics. Does it matter or not? Maybe it's fine for the vast majority of URL metrics to come from a single user - the viewport implications for the same viewport shouldn't differ wildly. But I want to make sure we consider that before going the route. Because effectively, for sites with very few to no users, implementing this change might lead to a similar result as if we had e.g. some admin UI to automatically visit all URLs as the authenticated user (which we discussed before, but never went for).

I didn't even think that far - so currently, this PR would result in an administrator being able to send all URL metrics necessary to fill a group? That would indeed be undesirable then, since e.g. refreshing the page a few times would already result in an entirely populated group (based on sample size). I guess I hadn't thought of that scenario yet, but that makes the concern even greater.

> Nevertheless, here's an alternative approach extending on top of what this PR currently implements. What if we use `sessionStorage` to keep track of the URLs which the client had submitted URL Metrics for? When `detect.js` boots up it can check if the `currentUrl` was already submitted, and if so, short-circuit. This addresses the privacy concern of storing some identifier about the user session with the URL Metric while also ensuring a diversity of clients being the source for the submitted URL Metrics. SG?

> There, what do you think about a57e080? The hashing of the URL in the `sessionStorage` may be overkill.

This would address the above concern right? Looks like a great solution. I think it's okay that a single administrator can provide a single metric per group per URL, so with this additional commit in place this should be a good approach.

westonruter commented:
Yes, the current state of the PR prevents any client (either an admin or not) from filling up the sample of URL Metrics. Previously, an anonymous user could have done so by waiting 60 seconds between page reloads, and initially this PR allowed an admin to do so by reloading multiple times even without waiting. Now only one URL Metric will be submitted for both. The difference is that when navigating on to a second URL, the admin will be able to submit a URL Metric for that one right away, whereas for the anonymous user they'll have to wait 60 seconds.

westonruter commented on plugins/optimization-detective/detect.js:
```diff
@@ -96,33 +96,65 @@ function error( ...message ) {
+				viewportWidth <= urlMetricGroupStatus.maximumViewportWidth )
 		) {
-			return complete;
+			return urlMetricGroupStatus;
```
There will be a merge conflict once 4c01373ff from #1833 is merged into `trunk`. That upstream change can be ignored since this PR implements the correct logic with the refactor.

westonruter committed:
Merge branch 'trunk' into remove/storage-lock-for-admins

felixarntz commented on plugins/optimization-detective/storage/class-od-storage-lock.php:
```diff
@@ -20,6 +20,49 @@
+		}
+
+		$primitive_cap = 'manage_options';
+		if ( self::STORE_URL_METRIC_NOW_CAPABILITY === $cap && user_can( $user_id, $primitive_cap ) ) {
```
The second check here is unnecessary, as that is checked after running the `map_meta_cap` filter anyway.
```suggestion
		if ( self::STORE_URL_METRIC_NOW_CAPABILITY === $cap ) {
```

Any `map_meta_cap` callback should return the list of capabilities required, and then Core will check that the user has all of them. So we don't need to check the same thing here too prior to that.

We can then also get rid of the `$user_id` parameter that is no longer used.

westonruter replied:
Good point. Done in bbba00163b8d949b00c9553484dcd606757cf62e.

felixarntz reviewed (commented):
@westonruter This looks good to me now, just one small note that would be good to address. It's doing a bit too much in the `map_meta_cap` callback.

westonruter committed:
Remove unnecessary checks from map_meta_cap

Co-authored-by: felixarntz <flixos90@git.wordpress.org>

westonruter committed:
Account for additional caps being required

westonruter commented:
@felixarntz I added one more improvement in 78a266e19e91c144bff8c461a4034cb2fb98daf3 which accounts for the case where another plugin may add a secondary capability to require for users who can `od_store_url_metric_now`. This additional commit will prevent clobbering any such additional caps.

westonruter commented on plugins/optimization-detective/detect.js:
```diff
@@ -308,14 +344,55 @@ export default async function detect( {
 	}
 
+	// Abort if the client already submitted a URL Metric for this URL and viewport group.
+	// TODO: Remove the console timing.
+	console.time( 'sha1' ); // eslint-disable-line no-console
+	const alreadySubmittedSessionStorageKey =
+		await getAlreadySubmittedSessionStorageKey(
+			currentETag,
+			currentUrl,
+			urlMetricGroupStatus
+		);
+	console.timeEnd( 'sha1' ); // eslint-disable-line no-console
```
@felixarntz What are your thoughts on the hashing of the key? The timing shows that this takes 10ms. Nevertheless, it is not blocking the main thread from. With 20x CPU slowdown it takes 35 ms:

> ![image](https://github.com/user-attachments/assets/842cab50-f0b8-4e8e-bf2f-b3d8476921d2)

See this trace: https://trace.cafe/t/CNJQYk4RTE

I also tried computing the hash 100 times in this function to better confirm that it is not blocking the main thread: https://trace.cafe/t/mBlNPnrYft

> ![image](https://github.com/user-attachments/assets/cf2a4bce-a515-4fa0-88d6-c41c73e51a7c)

So if this seems good to you and that hashing is warranted, then this suggestion can be accepted:

```suggestion
	const alreadySubmittedSessionStorageKey =
		await getAlreadySubmittedSessionStorageKey(
			currentETag,
			currentUrl,
			urlMetricGroupStatus
		);
```

felixarntz replied:
Thanks for the detailed research, that works for me! 👍

felixarntz commented:
> @felixarntz I added one more improvement in [78a266e](https://github.com/WordPress/performance/commit/78a266e19e91c144bff8c461a4034cb2fb98daf3) which accounts for the case where another plugin may add a secondary capability to require for users who can `od_store_url_metric_now`. This additional commit will prevent clobbering any such additional caps.

I don't think we should do that: Since `od_store_url_metric_now` is a meta capability based on the current implementation, other developers are not able to grant their own capabilities based on that capability.

If we want to make that possible, we need to instead handle `od_store_url_metric_now` as a base capability. That can be done via the `user_has_cap` filter (instead of `map_meta_cap`). Come to think of it, that's probably actually more appropriate for this capability, since there is nothing "meta" about it (e.g. there's no extra parameter to provide that the capability would depend on).

If you agree with that, we could simply solve it like this:
```php
add_filter(
	'user_has_cap',
	function ( $allcaps ) {
		if ( isset( $allcaps['manage_options'] ) ) {
			$allcaps['od_store_url_metric_now'] = $allcaps['manage_options'];
		}
		return $allcaps;
	}
);
```

FWIW, even WordPress Core has historically been doing it wrong for a bunch of cases - for some reason, `map_meta_cap` is widely referenced as the way to do it, but in many cases `user_has_cap` would actually be more appropriate.

westonruter committed:
Remove console timing

westonruter committed:
Revert "Account for additional caps being required"

This reverts commit 78a266e19e91c144bff8c461a4034cb2fb98daf3.

westonruter commented:
> I don't think we should do that: Since `od_store_url_metric_now` is a meta capability based on the current implementation, other developers are not able to grant their own capabilities based on that capability.

I'm not sure I understand. A meta capability maps to one or more primitive capabilities. Developers can add an arbitrary number of additional primitive capabilities to be required for a given meta capability. For example, to prevent a user from being able to do something, a developer can just push the `do_not_allow` capability into the list of primitive caps that are returned by `map_meta_cap()`. For `current_user_can()`, the user must have all of the capabilities returned by `map_meta_cap()`.

In any case, I reverted my commit. 

> Come to think of it, that's probably actually more appropriate for this capability, since there is nothing "meta" about it (e.g. there's no extra parameter to provide that the capability would depend on).

I don't think that the presence of args is really necessary for `map_meta_cap` to be used. For example, `edit_categories` and `delete_post_tags` always map to `manage_categories`. The `customize` cap always maps to `edit_theme_options`. 

**Update:** After discussing offline, I am largely won over and I've replaced `map_meta_cap` with `user_has_cap`.

westonruter committed:
Use user_has_cap instead of map_meta_cap

Co-authored-by: felixarntz <flixos90@git.wordpress.org>

felixarntz commented:
@westonruter We chatted about this via DM, but replying here too for visibility:

> > I don't think we should do that: Since `od_store_url_metric_now` is a meta capability based on the current implementation, other developers are not able to grant their own capabilities based on that capability.
> 
> I'm not sure I understand. A meta capability maps to one or more primitive capabilities. Developers can add an arbitrary number of additional primitive capabilities to be required for a given meta capability.

Yes, but if a developer wanted to rely their own meta capability on `od_store_url_metric_now` granted via `map_meta_cap`, they would need to make a nested call to `map_meta_cap()`, which is unnecessarily complex and slow. They only don't have to do it if we include a workaround like you had in the extra commit, but that's also unnecessarily complex - and almost no plugin does that, it's not something safe to rely on.

> > Come to think of it, that's probably actually more appropriate for this capability, since there is nothing "meta" about it (e.g. there's no extra parameter to provide that the capability would depend on).
> 
> I don't think that the presence of args is really necessary for `map_meta_cap` to be used. For example, `edit_categories` and `delete_post_tags` always map to `manage_categories`. The `customize` cap always maps to `edit_theme_options`.

All of these capabilities should really be primitive capabilities. They were probably made to use `map_meta_cap`, since almost all documentation only focuses on `map_meta_cap`, when in reality many capabilities would be better suited for `user_has_cap`. Consider for instance `edit_categories`: It would not be unreasonable if there was a `edit_category` meta capability expecting a term ID. Mapping that to `edit_categories` would be unnecessarily complicated the way it's implemented in Core today.

Obviously most of this can be worked around, but distinguishing between primitive and meta capabilities helps other developers using a plugin to be able to finetune how the (primitive) capabilities are granted.

felixarntz commented on plugins/optimization-detective/storage/class-od-storage-lock.php:
```diff
@@ -20,6 +20,44 @@
+	 * @return array<string, bool> Capability names mapped to boolean values for whether the user has that capability.
+	 */
+	public static function filter_user_has_cap( $allcaps ): array {
+		if ( ! is_array( $allcaps ) ) {
+			$allcaps = array();
+		}
```
Fair point to include this, although technically someone changing this to not be an array would be `doing_it_wrong`.

It's fine to keep this, but if we already cater for it, I wonder whether we should instead just return the same value in such a scenario. It's basically broken, and we know the condition below won't be satisfied anyway.

If some code does something unexpected, I'd rather just ignore it and pass through, than tamper with what they did wrong and possibly cause _more_ unexpected issues. :)

westonruter replied:
We do this in many other places actually. It's all for the sake of type safety, as otherwise a fatal error would occur since the function returns an `array`. This scenario happens when another plugin incorrectly returns the wrong type on a filter callback.

felixarntz replied:
I meant it mostly as: If someone else returns something other than an array, we could also return that same thing. The documented filter input type is basically always the same as the documented filter output type, so from that perspective we could also change the return type here to `array<string, bool>|mixed` and pass through any value that's not `array<string, bool>`.

westonruter replied:
Yeah, although that would require removing the PHP type hint from the function. All of the other filter callbacks in the repo have taken this approach, so if we want to pass through the value unchanged, then all the other callbacks should be updated. 

In any case, this is a safeguard for an edge condition for the sake of type safety.

felixarntz reviewed (approved):
@westonruter Thanks for bearing with me through the discussions and clarifying your points, this looks good to me now.

Just one non-blocking comment.

westonruter committed:
Update docs
